File: Server/src/main/content/global/skill/production/crafting/item/SnakeskinCraftPulse.java
package content.global.skill.production.crafting.item;

import content.global.skill.production.crafting.data.LeatherData;
import content.global.skill.production.crafting.data.SnakeskinData;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.SkillPulse;
import core.game.node.entity.skill.Skills;
import core.game.node.item.Item;
import core.game.world.update.flag.context.Animation;

public final class SnakeskinCraftPulse extends SkillPulse<Item> {

    private static final Animation ANIMATION = Animation.create(1249);
    private final SnakeskinData skin;
    private int amount;
    private int ticks;

    public SnakeskinCraftPulse(Player player, Item node, int amount, SnakeskinData skin) {
        super(player, node);
        this.amount = amount;
        this.skin = skin;
    }

    @Override
    public boolean checkRequirements() {
        if (player.getSkills().getLevel(Skills.CRAFTING) < skin.getLevel()) {
            player.getDialogueInterpreter().sendDialogue("You need a crafting level of " + skin.getLevel() + " to make this.");
            return false;
        }
        if (!player.getInventory().contains(LeatherData.NEEDLE, 1)) {
            return false;
        }
        if (!player.getInventory().containsItem(LeatherData.THREAD)) {
            player.getDialogueInterpreter().sendDialogue("You need thread to make this.");
            return false;
        }
        if (!player.getInventory().contains(6289, skin.getRequiredAmount())) {
            player.getDialogueInterpreter().sendDialogue("You need " + skin.getRequiredAmount() + " snakeskins in order to do this.");
            return false;
        }
        player.getInterfaceManager().close();
        return true;
    }

    @Override
    public void animate() {
        if (ticks % 5 == 0) {
            player.animate(ANIMATION);
        }
    }

    @Override
    public boolean reward() {
        if (++ticks % 5 != 0) {
            return false;
        }
        if (player.getInventory().remove(new Item(6289, skin.getRequiredAmount()))) {
            Item item = skin.getProduct();
            player.getInventory().add(item);
            player.getSkills().addExperience(Skills.CRAFTING, skin.getExperience(), true);
            LeatherData.decayThread(player);
            if (LeatherData.isLastThread(player)) {
                LeatherData.removeThread(player);
            }
        }
        amount--;
        return amount < 1;
    }

}

File: Server/src/main/content/global/skill/production/crafting/item/SnakeskinCraftPulse.kt
package content.global.skill.production.crafting.item

import content.global.skill.production.crafting.data.Leather
import content.global.skill.production.crafting.data.Leather.decayThread
import content.global.skill.production.crafting.data.Leather.isLastThread
import content.global.skill.production.crafting.data.Leather.removeThread
import content.global.skill.production.crafting.data.Snakeskin
import core.api.*
import cfg.consts.Animations
import cfg.consts.Items
import core.game.node.entity.player.Player
import core.game.node.entity.skill.SkillPulse
import core.game.node.entity.skill.Skills
import core.game.node.item.Item

/**
 * Snakeskin craft pulse
 *
 * @param amount The number of skins to be crafted.
 * @param skin The specific type of snakeskin being used.
 * @constructor Represents a new instance of SnakeskinCraftPulse.
 *
 * @param player The player who is performing the crafting action.
 * @param node The item node associated with the crafting process.
 */
class SnakeskinCraftPulse(player: Player?, node: Item?, var amount: Int, val skin: Snakeskin) : SkillPulse<Item?>(player, node) {

    var ticks = 0

    override fun checkRequirements(): Boolean {
        if (getStatLevel(player, Skills.CRAFTING) < skin.level) {
            sendDialogue(player, "You need a crafting level of " + skin.level + " to make this.")
            return false
        }
        if (!inInventory(player, Leather.NEEDLE, 1)) {
            return false
        }
        if (!inInventory(player, Leather.THREAD.id)) {
            sendDialogue(player, "You need thread to make this.")
            return false
        }
        if (!inInventory(player, Items.SNAKESKIN_6289, skin.requiredAmount)) {
            sendDialogue(player, "You need " + skin.requiredAmount + " snakeskins in order to do this.")
            return false
        }
        closeInterface(player)
        return true
    }

    override fun animate() {
        if (ticks % 5 == 0) {
            animate(player, ANIMATION)
        }
    }

    override fun reward(): Boolean {
        if (++ticks % 5 != 0) {
            return false
        }
        if (removeItem(player, Item(Items.SNAKESKIN_6289, skin.requiredAmount))) {
            val item = skin.product
            addItem(player, item.id)
            rewardXP(player, Skills.CRAFTING, skin.experience)
            decayThread(player)
            if (isLastThread(player)) {
                removeThread(player)
            }
        }
        amount--
        return amount < 1
    }

    companion object {
        private const val ANIMATION = Animations.CRAFT_LEATHER_1249
    }
}
